config.org
#+TITLE: Doom Emacs Configuration
#+AUTHOR: Robin Beckmann
#+EMAIL: robin.m.beckmann@gmail.com
#+PROPERTY: header-args :tangle yes


* General Settings
Set the users name and email address

#+BEGIN_SRC emacs-lisp
(setq user-full-name "Robin Beckmann"
      user-mail-address "robin.m.beckmann@gmail.com")
#+END_SRC
* Interface
Set the font and configure the line numbers correctly.
If set to ~'relative~, the numbering gets fucked up in org-mode
#+BEGIN_SRC emacs-lisp
(setq-default doom-font (font-spec :family "Hack" :size 14)
              display-line-numbers-type 'visual)
#+END_SRC

Show trailing whitespace.
#+BEGIN_SRC emacs-lisp
(setq show-trailing-whitespace t)
#+END_SRC

** Treemacs
*** Show current file in tree
Ever wondered where you are in your filetree? Never again, 'cause treemacs will
now follow your current file.
#+BEGIN_SRC emacs-lisp
(after! treemacs
  (treemacs-follow-mode))
#+END_SRC

*** Fix the ugly buttons
#+BEGIN_SRC emacs-lisp
(setq-default doom-themes-treemacs-theme "doom-colors")
#+END_SRC

* Completion Settings
** Projectile
Include hidden files in projectile's search
#+BEGIN_SRC emacs-lisp
(after! projectile
  (setq projectile-generic-command "find -L . -type f -print0"))
#+END_SRC

** Company
Reduce the minimum length of a sequence before company begins to display the
possible completions.
#+BEGIN_SRC emacs-lisp
(after! company
  (setq-default company-minimum-prefix-length 1))
#+END_SRC

Set ~company-quickhelp-delay~ to enable company-quickhelp and display it fairly
quickly.
#+BEGIN_SRC emacs-lisp
(after! company-quickhelp
  (setq company-quickhelp-delay 0.2))
#+END_SRC

** LSP
*** Elixir
Correct the language server executable...

#+BEGIN_SRC emacs-lisp
(setq lsp-clients-elixir-server-executable "elixir-ls")
#+END_SRC

* Eshell
** Aliases
- la :: ~ls -Al~
- ll :: ~ls -l~

#+BEGIN_SRC emacs-lisp
(after! eshell
  (set-eshell-alias!
   "la" "ls -lA"
   "ll" "ls -l"))
#+END_SRC
* Language specific settings
** Javascript
Set the correct javascript linter, because personally I prefer semicolons in
Javascript.
Furthermore Javascript code be indented using 2 spaces.
#+BEGIN_SRC emacs-lisp
(after! js2-mode
  (setq-default flycheck-javascript-standard-executable "semistandard"
                js2-basic-offset 2
                js-indent-level 2)
  (setq js2-basic-offset 2
        js-indent-level 2))
#+END_SRC

** Java
Java code should be indented using 4 spaces.
#+BEGIN_SRC emacs-lisp
(after! java-mode
  (setq-default c-basic-offset 4))
#+END_SRC

** PHP
Disable icons in the autocomplete popup for PHP, because it's buggy.
#+BEGIN_SRC emacs-lisp
(setq company-box-icons-acphp '(nil))
#+END_SRC

Get company to display completion for stdlib functions.
#+BEGIN_SRC emacs-lisp
(after! php-mode
  (set-company-backend! 'php-mode '(company-phpactor :seperate php-extras-company) 'company-dabbrev-code))
#+END_SRC

Enable phpactors hover function for eldoc
#+BEGIN_SRC emacs-lisp
(add-hook 'php-mode-hook
          (lambda ()
            (make-local-variable 'eldoc-documentation-function)
            (setq eldoc-documentation-function 'phpactor-hover)))
#+END_SRC

** Tex
*** Keybindings
- v :: View the current line in the generated PDF file.
- c :: Execute a TeX command on the master of the current file. Useful to
     rebuild the document.
#+BEGIN_SRC emacs-lisp
(after! latex
  (map! :map LaTeX-mode-map
        :localleader
        (:desc "View line" :n "v" #'TeX-view)
        (:desc "Command on master" :n "c" #'TeX-command-master)))
#+END_SRC

* Org-mode
Enable line numbers
#+BEGIN_SRC emacs-lisp
(remove-hook 'org-mode-hook #'doom|disable-line-numbers)
#+END_SRC

** Export settings
Enable the latex backend
#+BEGIN_SRC emacs-lisp
(after! org
  (add-to-list 'org-export-backends 'latex))
#+END_SRC

More beautiful html export
#+BEGIN_SRC emacs-lisp
(after! org
  (setq org-html-html5-fancy t
        org-html-doctype "html5"))
#+END_SRC

*** LaTeX Configuration
Add KomaScript documentclasses
#+BEGIN_SRC emacs-lisp
(after! ox-latex
  (add-to-list 'org-latex-classes
               '("koma-article"
                 "\\documentclass[11pt]{scrartcl}"
                 ("\\section{%s}" . "\\section*{%s}")
                 ("\\subsection{%s}" . "\\subsection*{%s}")
                 ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
                 ("\\paragraph{%s}" . "\\paragraph*{%s}")
                 ("\\subparagraph{%s}" . "\\subparagraph*{%s}")))
  (add-to-list 'org-latex-classes
               '("koma-report"
                 "\\documentclass[11pt]{scrreprt}"
                 ("\\part{%s}" . "\\part*{%s}")
                 ("\\chapter{%s}" . "\\chapter*{%s}")
                 ("\\section{%s}" . "\\section*{%s}")
                 ("\\subsection{%s}" . "\\subsection*{%s}")
                 ("\\subsubsection{%s}" . "\\subsubsection*{%s}")))
  (add-to-list 'org-latex-classes
               '("koma-book"
                 "\\documentclass[11pt]{scrbook}"
                 ("\\part{%s}" . "\\part*{%s}")
                 ("\\chapter{%s}" . "\\chapter*{%s}")
                 ("\\section{%s}" . "\\section*{%s}")
                 ("\\subsection{%s}" . "\\subsection*{%s}")
                 ("\\subsubsection{%s}" . "\\subsubsection*{%s}")))
  (setq org-latex-default-class "koma-article"))
#+END_SRC

** Keybindings
Activate all evil-org bindings.

#+BEGIN_SRC emacs-lisp
(after! org
  (setq evil-org-key-theme '(navigation
                             insert
                             return
                             textobjects
                             heading
                             shift
                             todo
                             additional
                             calendar)))
#+END_SRC

Add the following keybindings under the localleader
- e :: Invoke org-export. Easily export into bazillion different formats.
#+BEGIN_SRC emacs-lisp
(after! org
  (map! :map org-mode-map
        :localleader
        (:desc "Export" :n "e" #'org-export-dispatch)))
#+END_SRC
* Email (notmuch)
Set the backend to mbsync
#+BEGIN_SRC emacs-lisp
(setq-default +notmuch-sync-backend 'mbsync)
#+END_SRC

I want to see my modeline -__-
#+BEGIN_SRC emacs-lisp
(after! notmuch
  (setq notmuch-message-headers-visible t)
  (remove-hook! '(notmuch-show-mode-hook
                  notmuch-tree-mode-hook
                  notmuch-search-mode-hook)
                #'hide-mode-line-mode)
)
#+END_SRC


config.el
(setq user-full-name "Robin Beckmann"
      user-mail-address "robin.m.beckmann@gmail.com")

(setq-default doom-font (font-spec :family "Hack" :size 14)
              display-line-numbers-type 'visual)

(setq show-trailing-whitespace t)

(after! treemacs
  (treemacs-follow-mode))

(after! projectile
  (setq projectile-generic-command "find -L . -type f -print0"))

(after! company
  (setq-default company-minimum-prefix-length 1))

(after! company-quickhelp
  (setq company-quickhelp-delay 0.2))

(after! eshell
  (set-eshell-alias!
   "la" "ls -lA"
   "ll" "ls -l"))

(after! js2-mode
  (setq-default flycheck-javascript-standard-executable "semistandard"
                js2-basic-offset 2
                js-indent-level 2)
  (setq js2-basic-offset 2
        js-indent-level 2))

(after! java-mode
  (setq-default c-basic-offset 4))

(setq company-box-icons-acphp '(nil))

(add-hook 'php-mode-hook
          (lambda ()
            (make-local-variable 'eldoc-documentation-function)
            (setq eldoc-documentation-function 'phpactor-hover)))

(after! latex
  (map! :map LaTeX-mode-map
        :localleader
        (:desc "View line" :n "v" #'TeX-view)
        (:desc "Command on master" :n "c" #'TeX-command-master)))

(remove-hook 'org-mode-hook #'doom|disable-line-numbers)

(after! org
  (add-to-list 'org-export-backends 'latex))

(after! org
  (setq org-html-html5-fancy t
        org-html-doctype "html5"))

(after! ox-latex
  (add-to-list 'org-latex-classes
               '("koma-article"
                 "\\documentclass[11pt]{scrartcl}"
                 ("\\section{%s}" . "\\section*{%s}")
                 ("\\subsection{%s}" . "\\subsection*{%s}")
                 ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
                 ("\\paragraph{%s}" . "\\paragraph*{%s}")
                 ("\\subparagraph{%s}" . "\\subparagraph*{%s}")))
  (add-to-list 'org-latex-classes
               '("koma-report"
                 "\\documentclass[11pt]{scrreprt}"
                 ("\\part{%s}" . "\\part*{%s}")
                 ("\\chapter{%s}" . "\\chapter*{%s}")
                 ("\\section{%s}" . "\\section*{%s}")
                 ("\\subsection{%s}" . "\\subsection*{%s}")
                 ("\\subsubsection{%s}" . "\\subsubsection*{%s}")))
  (add-to-list 'org-latex-classes
               '("koma-book"
                 "\\documentclass[11pt]{scrbook}"
                 ("\\part{%s}" . "\\part*{%s}")
                 ("\\chapter{%s}" . "\\chapter*{%s}")
                 ("\\section{%s}" . "\\section*{%s}")
                 ("\\subsection{%s}" . "\\subsection*{%s}")
                 ("\\subsubsection{%s}" . "\\subsubsection*{%s}")))
  (setq org-latex-default-class "koma-article"))

(after! org
  (setq evil-org-key-theme '(navigation
                             insert
                             return
                             textobjects
                             heading
                             shift
                             todo
                             additional
                             calendar)))

(after! org
  (map! :map org-mode-map
        :localleader
        (:desc "Export" :n "e" #'org-export-dispatch)))
